### ISSUE TITLE
Race in a finalizer is not executed in case of interruption

### ISSUE THREAD

**Issue by vasilmkd** [zio/zio#2226]
Basically, what the title says.

Running the spec on a fresh sbt instance once runs fine. Running it a second time times out. Running it a third time, it goes into a `stacktracerJVM compileIncremental` loop and never executes again.

This is a test that usually takes less than 2 seconds to run. It involves interruption. I'm pretty sure it has something to do with #2169.

-- Edit by @iravid --

The symptom here is that the tests in that suite use the global pool to run blocking code. That global pool is eventually exhausted after a few runs, because we're not getting an interruption when offering to a shutdown queue.

-- Edit by @ghostdogpr 

Changed the title to describe the root problem which is that `race` inside a finalizer is not executed properly in case of interruption.

**Comment by vasilmkd** [comment]
What I described earlier happens on a 2 core, 4 threads machine. I tried on a 4 core, 8 threads machine and it happens after the third time the test is started.

**Comment by vasilmkd** [comment]
Here is a Java flight recording of the sbt instance running the tests with the above specified workflow. The StreamEffectAsyncSpec test was run once, then again, without issues. The third time it timed out and the fourth time it went into the `stracktracerJVM compileIncremental` loop. Again, this was after a clean compile cycle, on master, with no changes.

[flight_recording_180222xsbtbootBoot10508.zip](https://github.com/zio/zio/files/3841338/flight_recording_180222xsbtbootBoot10508.zip)

**Comment by iravid** [comment]
Yes, nice catch, this looks like a bug! I'm editing the description and title with a minimized reproduction.

**Comment by vasilmkd** [comment]
Funny that you just merged `fromQueue`, because it relies on this behavior.

**Comment by iravid** [comment]
Hmm, actually, I can't reproduce this with a test. So there might be something else happening here.

**Comment by iravid** [comment]
Sorry, I was fooled by `onInterrupt` which doesn't act on internal interruption. This seems to be a different issue.

**Comment by vasilmkd** [comment]
Sorry, I've been away for a couple of days, does that mean that the issue description is out of date as well, with regards to your last comment? Thanks.

**Comment by ghostdogpr** [comment]
We have a test to verifies that offers to a shutdown queue result in interruption:
```scala
 testM("shutdown with offer") {
      for {
        selfId <- ZIO.fiberId
        queue  <- Queue.bounded[Int](1)
        _      <- queue.shutdown
        res    <- queue.offer(1).sandbox.either
      } yield assert(res, isLeft(equalTo(Cause.interrupt(selfId))))
    }
```
This test pass properly so I'm wondering if that's the real problem?

**Comment by vasilmkd** [comment]
Putting a prints statement after the shutting down of the queue never gets printed like so:
```scala
output  <- Queue.bounded[Pull[R, E, A]](outputBuffer).toManaged(_.shutdown *> UIO.effectTotal(println("hello")))
```

The queue has trouble shutting down.

**Comment by ghostdogpr** [comment]
Do you have a way to reproduce that? If I execute this snippet, "hello" gets printed well.

**Comment by vasilmkd** [comment]
Run the `streamsTestsJVM/testOnly zio.stream.StreamEffectAsyncSpec -- -t "effectAsyncMaybe back pressure" or any of the other "back pressure" tests.

**Comment by ghostdogpr** [comment]
OK, I was able to reproduce in that particular test. When the queue shuts down, it calls `p.interruptAs(fiberId)` on each pending offer p (of type `Promise`), and somehow this call never returns.

**Comment by vasilmkd** [comment]
Exactly what I observed as well. Unfortunately, my knowledge of the new interruption system ends there. I haven't had the time to dive deeper these weeks.

**Comment by ghostdogpr** [comment]
cc @jdegoes 

I’ll keep digging as well.

**Comment by adamgfraser** [comment]
`interruptAs` is going to semantically block until the fiber and its children have been interrupted. So I would think conceptually the change is that previously just the `run` fiber had to be interrupted to continue, whereas now all its non-daemon children must be as well. So could there be a child fiber in `ZSink#fromEffect` that is not interruptible, and thus is causing waiting on termination to suspend indefinitely?

**Comment by vasilmkd** [comment]
`ZSink.fromEffect` just evaluates the passed effect, which in this case is `ZIO.never`.

**Comment by adamgfraser** [comment]
Right, sorry. But the `run` fiber is describing the entire evaluation of the stream with the sync. So all child fibers and their children recursively created in evaluating that stream have to be interrupted. Could we do a fiber dump to see what is happening?

**Comment by ghostdogpr** [comment]
Got off work, I’ll do it tomorrow morning unless someone figures it out before. A good occasion to try out fiber dumps!

**Comment by ghostdogpr** [comment]
The `p.interruptAs(fiberId)` I was talking about is actually never executed. It seems that the shutdown itself gets interrupted, even though it is called in a `Managed` release action, and it is even explicitly marked `.uninterruptible` in `ZQueue#shutdown`.

I added `onTermination` here:
```scala
final val shutdown: UIO[Unit] =
      ZIO.fiberId.flatMap(
        fiberId =>
          IO.whenM(shutdownHook.succeed(()))(
              IO.effectTotal(unsafePollAll(takers)) >>= (IO.foreachPar(_)(_.interruptAs(fiberId)) *> strategy.shutdown)
            )
            .uninterruptible
            .onTermination(cause => UIO(println(cause.prettyPrint)))
      )
```
And it prints that the fiber was interrupted.

I dug a bit more to find where the interruption came from and it seems it's from a `raceWith`. The recently added low level implementation of race has this:
```scala
val left  = fork[EL, A](race.left.interruptible.asInstanceOf[IO[EL, A]])
val right = fork[ER, B](race.right.interruptible.asInstanceOf[IO[ER, B]])
```
Shouldn't a mask be used here to preserve the interruptibility of the underlying code? If I brute-force remove `.interruptible`, the shutdown completes properly.

I'm not an expert of this part of the code, so let me know what you guys think.

**Comment by adamgfraser** [comment]
That looks right to me. `uninterruptibleMask` was used in the previous implementation.

https://github.com/zio/zio/pull/2091/files

**Comment by ghostdogpr** [comment]
OK I think I figured what happens more precisely:
1. the fiber gets interrupted (from the test)
2. the finalizer of `Managed` shuts down the queue
3. the shutdown operation runs a `foreachPar` to interrupt pending offers, resulting in a `race`
4. `race` creates 2 fibers and calls `.interruptible` on them (see my previous comment)
5. the created fibers have initially `isInterrupted = false` and `isInterruptible = false`
6. then for each of those 2 fibers, the function `propagateAncestorInterruption` gets the interrupted status from the parent and propagate it to the current fiber: `isInterrupted = true`
7. `.interruptible` is executed, changing `isInterruptible = true`
8. we now have `isInterrupted = true` and `isInterruptible = true`, processing stops and whatever was in the race is not executed

In other words, it seems that a race executed in a finalizer will not be ran in case of interruption.

**Comment by iravid** [comment]
@neko-kai looks like what you were talking about in the issue about interruption in finalizers...

**Comment by iravid** [comment]
Great work diagnosing this @ghostdogpr

**Comment by ghostdogpr** [comment]
Easy repro:
```scala
testM("race in finalizer is executed") {
  for {
    ref   <- Ref.make(0)
    fiber <- ZIO.unit.bracket(_ => ref.set(1) race ref.set(1))(_ => ZIO.never).fork
    _     <- fiber.interrupt
    ref   <- ref.get
  } yield assert(ref, equalTo(1))
}
```

**Comment by neko-kai** [comment]
@ghostdogpr @iravid 
> In other words, it seems that a race executed in a finalizer will not be ran in case of interruption.

Sounds similar to https://github.com/zio/zio/issues/2005

**Comment by ghostdogpr** [comment]
Isn't the issue with `propagateAncestorInterruption`? Having an interruptible zone within a finalizer seems like a possible case, but inheriting the `interrupted` state of the parent hierarchy will cause it to be interrupted.

@jdegoes what do you think?

**Comment by jdegoes** [comment]
`race` forces both left and right to be interruptible. Maybe we should re-investigate that.

`propagateAncestorInterruption` is just a shortcircuit enabling child fibers to shut down a little faster, and doesn't actually affect interruptibility.

**Comment by neko-kai** [comment]
@jdegoes 
> race forces both left and right to be interruptible. Maybe we should re-investigate that.

Not making them interruptible would just cause a deadlock when interrupting the losing side when `race` is used in finalizer, not an improvement from the bug in question. Or do you mean that race interruption should be special somehow, not subject to regions?

**Comment by ghostdogpr** [comment]
> `propagateAncestorInterruption` is just a shortcircuit enabling child fibers to shut down a little faster, and doesn't actually affect interruptibility.

@jdegoes It's doing more than that: it causes the interruption of children fibers that are forked by a finalizer. We don't want these to inherit the `interrupted` status of their parent. The parent would not attempt to kill those children. If I remove the call to `propagateAncestorInterruption`, the test pass well.

**Comment by vasilmkd** [comment]
Passes as in, it doesn't exhaust the global thread pool? @ghostdogpr

**Comment by ghostdogpr** [comment]
Yeah the finalizers aren’t getting cancelled so everything is cleaned up.

**Comment by ghostdogpr** [comment]
Summarized the problem and proposed fix: https://github.com/zio/zio/pull/2317

**Comment by ghostdogpr** [comment]
I ran my old reproducing test with RC18 and I can confirm it works properly.
Closing this ticket and big 👏 @jdegoes

**Comment by jdegoes** [comment]
@ghostdogpr Thanks for confirming!

**Comment by vasilmkd** [comment]
I can also confirm that the `streams` tests that were exhibiting this behavior now work correctly. Thanks!

**Comment by adamgfraser** [comment]
Awesome!
